simple/hex/h2.py

```diff
@@ -212,26 +220,35 @@ def sim_test(brd, uf_p, ptm, sims):
     if ptm == simulate(brd, uf_p, ptm):
       wins += 1
   print(Cell.ch[ptm], ' to play', wins, '/', sims, 'wins')
-    
+
 def mcts(board, uf_parents, root_ptm, max_iterations, expand_threshold):
-  def descend(node, board, ptm):
-    node = tree_policy_child(node, root_ptm - ptm)
-    putstone_and_update(board, uf_parents, node.move, ptm)
+  def descend(node, board, ptm, uf_par):
+    node = node.tree_policy_child(root_ptm - ptm)
+    putstone_and_update(board, uf_par, node.move, ptm)
     return node, board, Cell.opponent(ptm)
 
-  root_node, iterations = Mcts_node(None, None), 0
+  root_node, iterations = Mcts_node('root', None), 0
+  root_node.expand_node(board)
+  ptm = root_ptm
   while iterations < max_iterations:
     node, ptm = root_node, root_ptm
     brd, uf_par = deepcopy(board), deepcopy(uf_parents)
-    while not node.is_leaf():      # select leaf
-      node, brd, ptm = descend(node, brd, P, ptm)
+
+    while not node.is_leaf():            # select leaf
+      node, brd, ptm = descend(node, brd, ptm, uf_par)
+
     if node.visits > expand_threshold:
-      expand_node(node, brd)     # expand
-      node, brd, ptm = descend(node, brd, P, ptm)
-    result = simulate(brd, P, ptm)  # simulate
-    while node.has_parent():       # propagate
-      update(node, root_ptm, result)
+      node.expand_node(brd)              # expand
+      node, brd, ptm = descend(node, brd, ptm, uf_par)
+
+    result = simulate(brd, uf_par, ptm)  # simulate
+    while node.has_parent():             # propagate
```

Comment:
this loop is a bit awkward to read, and node.has_parent() seems to be called more often than needed;

maybe replace with loop with something of the form:
while True:
  ...
  and then have a break whenever you want to leave the loop, this way might be cleaner